src/crypto/CHIPCryptoPALmbedTLS.cpp

```diff
@@ -1250,14 +1250,73 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::PointIsValid(void * R)
+        auto valueA = timeA->CHIP_CRYPTO_PAL_PRIVATE_X509(component);                                                              \
+        auto valueB = timeB->CHIP_CRYPTO_PAL_PRIVATE_X509(component);                                                              \
+                                                                                                                                   \
+        if (valueA != valueB)                                                                                                      \
+        {                                                                                                                          \
+            return valueA > valueB;                                                                                                \
+        }                                                                                                                          \
+    }
+
+    RETURN_STRICTLY_GREATER_IF_DIFFERENT(year);
+    RETURN_STRICTLY_GREATER_IF_DIFFERENT(mon);
+    RETURN_STRICTLY_GREATER_IF_DIFFERENT(day);
+    RETURN_STRICTLY_GREATER_IF_DIFFERENT(hour);
+    RETURN_STRICTLY_GREATER_IF_DIFFERENT(min);
+    RETURN_STRICTLY_GREATER_IF_DIFFERENT(sec);
+
+    // all above are equal
+    return true;
+}
+
+CHIP_ERROR IsCertificateValidAtIssuance(const mbedtls_x509_crt * referenceCertificate,
+                                        const mbedtls_x509_crt * toBeEvaluatedCertificate)
+{
+    mbedtls_x509_time refNotBeforeTime = referenceCertificate->CHIP_CRYPTO_PAL_PRIVATE_X509(valid_from);
+    mbedtls_x509_time tbeNotBeforeTime = toBeEvaluatedCertificate->CHIP_CRYPTO_PAL_PRIVATE_X509(valid_from);
+    mbedtls_x509_time tbeNotAfterTime  = toBeEvaluatedCertificate->CHIP_CRYPTO_PAL_PRIVATE_X509(valid_to);
+
+    // check if referenceCertificate is issued at or after tbeCertificate's notBefore timestamp
+    VerifyOrReturnError(IsTimeGreaterThanEqual(&refNotBeforeTime, &tbeNotBeforeTime), CHIP_ERROR_CERT_EXPIRED);
+
+    // check if referenceCertificate is issued at or before tbeCertificate's notAfter timestamp
+    VerifyOrReturnError(IsTimeGreaterThanEqual(&tbeNotAfterTime, &refNotBeforeTime), CHIP_ERROR_CERT_EXPIRED);
+
+    return CHIP_NO_ERROR;
+}
+
+static int CallbackForCustomValidityCheck(void * data, mbedtls_x509_crt * crt, int depth, uint32_t * flags)
 {
+    mbedtls_x509_crt * leafCert = reinterpret_cast<mbedtls_x509_crt *>(data);
+    mbedtls_x509_crt * tbeCert  = crt;
+
+    // Ignore any time validy error performed by the standard mbedTLS code.
     *flags &= ~(static_cast<uint32_t>(MBEDTLS_X509_BADCERT_EXPIRED | MBEDTLS_X509_BADCERT_FUTURE));
+
+    // Verify that all certificates in the chain were valid at the time of
+    // the leaf certificate generation (notBefore time of the leaf certificate).
+    // Note that this callback is invoked for each certificate in the chain.
```

Comment:
We actually want to check that the leaf certificate has a notBefore valid during the validity period of issuer (e.g. a DAC against a PAI or PAA). The wording here sounds like the opposite.

Reply:
updated